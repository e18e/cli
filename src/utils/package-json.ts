import type {FileSystem} from '../file-system.js';
import type {PackageJsonLike} from '../types.js';

export async function getPackageJson(
  fileSystem: FileSystem
): Promise<PackageJsonLike | null> {
  let packageJsonText: string;

  try {
    packageJsonText = await fileSystem.readFile('/package.json');
  } catch {
    // No package.json found
    return null;
  }

  try {
    return JSON.parse(packageJsonText);
  } catch {
    // Not parseable
    return null;
  }
}
