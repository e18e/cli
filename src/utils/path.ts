export function normalizePath(path: string): string {
  return path.replaceAll('\\', '/');
}
