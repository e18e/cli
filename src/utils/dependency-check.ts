import {createRequire} from 'node:module';
import {spawn} from 'node:child_process';
import {detect, resolveCommand} from 'package-manager-detector';
import * as p from '@clack/prompts';
import c from 'picocolors';
import {resolve} from 'node:path';

const require = createRequire(import.meta.url);

/**
 * Checks if a package is installed and can be imported
 */
export function isPackageInstalled(packageName: string): boolean {
  try {
    // Check only in current directory's node_modules, not parent directories
    const localPath = resolve('./node_modules', packageName);
    require.resolve(localPath);
    return true;
  } catch {
    return false;
  }
}

/**
 * Prompts the user to install a missing package
 * Returns true if user wants to proceed with installation
 */
export async function promptToInstall(packageName: string): Promise<boolean> {
  p.log.warn(
    `${c.yellow('Optional dependency not found:')} ${c.bold(packageName)}`
  );
  p.log.message(`This package is required for the feature you've enabled.`, {
    spacing: 0
  });

  const shouldInstall = await p.confirm({
    message: `Would you like to install ${c.cyan(packageName)} now?`
  });

  if (p.isCancel(shouldInstall)) {
    return false;
  }

  return shouldInstall;
}

/**
 * Automatically installs a package using the detected package manager
 */
export async function installPackage(packageName: string): Promise<boolean> {
  const detected = await detect();
  const agent = detected?.agent || 'npm';

  const resolved = resolveCommand(agent, 'add', [packageName]);

  if (!resolved) {
    p.log.error(`Failed to resolve install command for ${c.cyan(agent)}`);
    return false;
  }

  p.log.info(`Installing ${c.cyan(packageName)} with ${c.cyan(agent)}...`);

  return new Promise((resolve) => {
    const child = spawn(resolved.command, resolved.args, {
      stdio: 'inherit',
      cwd: process.cwd()
    });

    child.on('close', (code) => {
      if (code === 0) {
        p.log.success(`Successfully installed ${c.cyan(packageName)}!`);
        resolve(true);
      } else {
        p.log.error(`Failed to install ${c.cyan(packageName)}`);
        resolve(false);
      }
    });
  });
}
