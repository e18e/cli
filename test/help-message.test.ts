import { execSync } from 'node:child_process';
import { join } from 'node:path';
import { describe, expect, it } from 'vitest';
import { stripVersion } from './utils.js';

describe('--help command', () => {
  it('prints the up to date help command', () => {
    const cliPath = join(__dirname, '..', 'cli.js');

    const output = execSync(`node ${cliPath} --help`, {
      stdio: 'pipe',
    });

    expect(stripVersion(output.toString())).toMatchSnapshot();
  });

  it('prints the analyze command help', () => {
    const cliPath = join(__dirname, '..', 'cli.js');

    const output = execSync(`node ${cliPath} analyze --help`, {
      stdio: 'pipe',
    });

    const cleanOutput = stripVersion(output.toString());
    expect(cleanOutput).toMatchSnapshot();
  });

  it('prints the migrate command help', () => {
    const cliPath = join(__dirname, '..', 'cli.js');

    const output = execSync(`node ${cliPath} migrate --help`, {
      stdio: 'pipe',
    });

    const cleanOutput = stripVersion(output.toString());
    expect(cleanOutput).toMatchSnapshot();
  });

  it('prints the up to date version', () => {
    const cliPath = join(__dirname, '..', 'cli.js');

    const output = execSync(`node ${cliPath} --version`, {
      stdio: 'pipe',
    });

    expect(output.toString().trim()).toMatch(/^\d+\.\d+\.\d+(?:-\S+)?$/); /** major.minor.patch[-prerelease] */
  });
});
