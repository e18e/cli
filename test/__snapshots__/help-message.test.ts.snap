// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`--help command > prints the analyze command help 1`] = `
"[36me18e[39m (cli <version>)

Analyze the project for any warnings or errors

USAGE:
  cli analyze <OPTIONS>

OPTIONS:
  --pack [pack]                    Package manager to use for packing (default: auto, choices: auto | npm | yarn | pnpm | bun | none)
  --log-level [log-level]          Set the log level (debug | info | warn | error) (default: info, choices: debug | info | warn | error)
  -h, --help                       Display this help message
  -v, --version                    Display this version

"
`;

exports[`--help command > prints the migrate command help 1`] = `
"[36me18e[39m (cli <version>)

Migrate from a package to a more performant alternative.

USAGE:
  cli migrate <OPTIONS>

OPTIONS:
  --dry-run                    Don't apply any fixes, only show what would change. (default: false)
  --interactive                Run in interactive mode. (default: false)
  --include [include]          Files to migrate (default: **/*.{ts,js})
  -h, --help                   Display this help message
  -v, --version                Display this version

"
`;

exports[`--help command > prints the up to date help command 1`] = `
"[36me18e[39m (cli <version>)

USAGE:
  cli [(anonymous)] <OPTIONS>
  cli <COMMANDS>

COMMANDS:
  analyze          Analyze the project for any warnings or errors  
  migrate          Migrate from a package to a more performant alternative.  

For more info, run any command with the \`--help\` flag:
  cli analyze --help
  cli migrate --help

OPTIONS:
  -h, --help             Display this help message
  -v, --version          Display this version

"
`;
